src/controller/CHIPDeviceController.cpp

```diff
@@ -2189,29 +2190,110 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
 }
 #endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
 
-// ClusterStateCache::Callback impl
+// ClusterStateCache::Callback / ReadClient::Callback
 void DeviceCommissioner::OnDone(app::ReadClient * readClient)
 {
     VerifyOrDie(readClient != nullptr && readClient == mReadClient.get());
     mReadClient.reset();
     switch (mCommissioningStage)
     {
     case CommissioningStage::kReadCommissioningInfo:
-        // Silently complete the stage, data will be saved in attribute cache and
-        // will be parsed after all ReadCommissioningInfo stages are completed.
-        CommissioningStageComplete(CHIP_NO_ERROR);
-        break;
-    case CommissioningStage::kReadCommissioningInfo2:
-        // Note: Only parse commissioning info in the last ReadCommissioningInfo stage.
-        ParseCommissioningInfo();
+        ContinueReadingCommissioningInfo(mCommissioningDelegate->GetCommissioningParameters());
         break;
     default:
         VerifyOrDie(false);
         break;
     }
 }
 
-void DeviceCommissioner::ParseCommissioningInfo()
+void DeviceCommissioner::ContinueReadingCommissioningInfo(const CommissioningParameters & params)
+{
+    VerifyOrDie(mCommissioningStage == CommissioningStage::kReadCommissioningInfo);
+
+    // We can ony read 9 paths per Read Interaction, since that is the minimum
+    // a server has to support per spec (see "Interaction Model Limits").
+    app::AttributePathParams readPaths[InteractionModelEngine::kMinSupportedPathsPerReadRequest];
+    static_assert(InteractionModelEngine::kMinSupportedPathsPerReadRequest >= 9);
+
+    const auto proxy    = mDeviceBeingCommissioned;
```

Comment:
So what happens if we were in the middle of a read and commissioning got canceled?  I think proxy here will be null, right?  

In general, the handling of things that happen during the async parts of this process needs some careful thinking about.  It was handled by the general "go to next step" machinery before, but now we have to make sure it's correct.

Reply:
Hm, that was true for the initial `kReadCommissioningInfo` step, but I'm not sure cancellation during `kRead...2` was actually handled correctly. In `CancelCommissioningInteractions` we kill the ReadClient, and this calls `OnError` and `OnDone`. Only `OnDone` is currently implemented, and `ParseCommissioningInfo` will then probably happen to fail... Maybe `CancelCommissioningInteractions` needs to just finish the step first, and then we can just ignore the `OnDone` call.

Reply:
Actually that's not quite correct... the ReadClient destructor won't call `OnDone`, so I don't think there is anything different to be done here about how we handle cancellation before and after my change.

Reply:
Just to elaborate a bit more, cancellation calls `CancelCommissioningInteractions()`, we free the ReadClient (and now also the ClusterStateCache, this was missing before), and `OnDone` won't be called. So we won't get back into `ContinueReadingCommissioningInfo` or touch `mDeviceBeingCommissioned` in that case.